src/System.Management.Automation/engine/Modules/AnalysisCache.cs

```diff
@@ -1025,12 +1027,52 @@ private AnalysisCacheData()
 
         static AnalysisCacheData()
         {
-            s_cacheStoreLocation =
-                Environment.GetEnvironmentVariable("PSModuleAnalysisCachePath") ??
+            // If user defines a custom cache path, then use that.
+            string userDefinedCachePath = Environment.GetEnvironmentVariable("PSModuleAnalysisCachePath");
+            if (!String.IsNullOrEmpty(userDefinedCachePath))
+            {
+                s_cacheStoreLocation = userDefinedCachePath;
+            }
+
+            string cacheFileName = "ModuleAnalysisCache";
+            if (ExperimentalFeature.EnabledExperimentalFeatureNames.Count > 0)
+            {
+                // If any experimental features are enabled, we cannot use the default cache file because those
+                // features may expose commands that are not available in a regular powershell session, and we
+                // should not cache those commands in the default cache file because that will result in wrong
+                // auto-completion suggestions when the default cache file is used in another powershell session.
+                //
+                // Here we will generate a cache file name that represent the combination of enabled feature names.
+                // We first convert enabled feature names to lower case, then we sort the feature names, and then
+                // compute an SHA1 hash from the sorted feature names. We will use a short SHA name (first 8 chars)
+                // to generate the cache file name.
+                int index = 0;
+                string[] featureNames = new string[ExperimentalFeature.EnabledExperimentalFeatureNames.Count];
+                foreach (var featureName in ExperimentalFeature.EnabledExperimentalFeatureNames)
+                {
+                    featureNames[index++] = featureName.ToLowerInvariant();
+                }
+
+                Array.Sort(featureNames);
+                string allNames = String.Join(Environment.NewLine, featureNames);
```

Comment:
maybe it's better to join with a static value instead of NewLine?  Is NewLine static or based on culture?

Reply:
I think newline is platform based: `\n` vs `\r\n`

Reply:
Which I think is static on a given platform?

Reply:
I looked into it, `Environment.NewLine` is [compile-time static](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L312) (the way corefx does it is they have a partial class with the shared part and then a `.Unix.cs` and a `.Windows.cs`, etc. and then they [conditionally include it in compilation](https://github.com/dotnet/corefx/blob/af8f03ce8b0c6e16cb28dda27d637356fde4c99b/src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj#L181-L182)).

Reply:
Yes, `Environment.NewLine` is static, `\r\n` on Windows and `\n` on Unix platforms. Here we are calculating a hash that is used locally for the analysis cache file name, so it's OK as long as it's consistent on a specific platform.